compiler/codegen/CCData.cpp

```diff
@@ -0,0 +1,144 @@
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "omrcomp.h"
+#include "OMR/Bytes.hpp"
+
+#include "codegen/CCData.hpp"
+
+#include "infra/Monitor.hpp"
+#include "infra/CriticalSection.hpp"
+
+using OMR::CCData;
+
+CCData::key_t CCData::key(const uint8_t * const data, const size_t sizeBytes)
+   {
+   return key_t(reinterpret_cast<const key_t::value_type *>(data), sizeBytes);
+   }
+
+CCData::key_t CCData::key(const char * const str)
+   {
+   return key_t(reinterpret_cast<const key_t::value_type *>(str));
+   }
+
+// Converts a size in units of bytes to a size in units of sizeof(data_t).
+#define DATA_SIZE_FROM_BYTES_SIZE(x) (((x) + (sizeof(data_t)) - 1) / (sizeof(data_t)))
+
+// Converts an alignment in units of bytes to an alignment in units of alignof(data_t).
+#define DATA_ALIGNMENT_FROM_BYTES_ALIGNMENT(x) (((x) + (OMR_ALIGNOF(data_t)) - 1) / (OMR_ALIGNOF(data_t)))
+
+CCData::CCData(const size_t sizeBytes)
+: _data(new data_t[DATA_SIZE_FROM_BYTES_SIZE(sizeBytes)]), _capacity(DATA_SIZE_FROM_BYTES_SIZE(sizeBytes)), _putIndex(0), _lock(TR::Monitor::create("CCDataMutex")), _releaseData(true)
+   {
+   }
+
+CCData::CCData(uint8_t * const storage, const size_t sizeBytes)
+: _putIndex(0), _lock(TR::Monitor::create("CCDataMutex")), _releaseData(false)
+   {
+   void *alignedStorage = storage;
+   size_t sizeBytesAfterAlignment = sizeBytes;
+   bool success = OMR::align(OMR_ALIGNOF(data_t), sizeof(data_t), alignedStorage, sizeBytesAfterAlignment) != NULL;
+   TR_ASSERT_FATAL(success, "Can't align CCData storage to required boundary");
+   _data = reinterpret_cast<data_t *>(alignedStorage);
+   _capacity = DATA_SIZE_FROM_BYTES_SIZE(sizeBytesAfterAlignment);
+   }
+
+CCData::~CCData()
+   {
+   // Memory for data can either be allocated by this class or passed in via
+   // the constructor. If allocated, it has to be freed now; if passed in we can't free
+   // it.
+   if (_releaseData)
+      delete [] _data;
```

Comment:
This seems like an anty pattern because it's conflating the management of the memory for `_data` with the implementation of CCData. It would be good to separate these two distinct concepts.

Reply:
Agreed. See https://github.com/eclipse/omr/pull/5634/files#r522987673.